### PR TITLE
3D Visualization Upgrade

### PR DESCRIPTION
**Pull request by nontas**
This PR exposes various rendering options in the `_view_3d()` and `_view_landmarks_3d()` methods of Menpo's objects. 3D visualization is now equally powerful to 2D and follows the same logic.

This PR depends on https://github.com/menpo/menpo3d/pull/36 .

Note that the only major difference to the 2D case is the fact that the default value of `new_figure` argument is set to ``True``.

### CODE REVIEW

File: menpo/shape/mesh/textured.py [hunk 1]
```diff
@@ -218,39 +227,142 @@ def from_mask(self, mask):
...
+        range : ``(float, float)``, optional
+            The clipping range.
+
+        Returns
+        -------
+        self : :map:`ColouredTriMesh`
+            A copy of self with its texture clipped.
+        """
+        instance = self.copy()
+        instance.texture.pixels = np.clip(self.texture.pixels, *range)
```
Comment (patricksnape) [inline menpo/shape/mesh/textured.py]: Maybe we should put a clip method on our images?

File: menpo/shape/pointcloud.py [hunk 2]
```diff
@@ -777,49 +780,186 @@ def _view_3d(self, figure_id=None, new_figure=False):
...
         try:
-            from menpo3d.visualize import PointCloudViewer3d
-            return PointCloudViewer3d(figure_id, new_figure,
-                                      self.points).render()
+            from menpo3d.visualize import PointGraphViewer3d
+            edges = np.empty(0)
+            renderer = PointGraphViewer3d(figure_id, new_figure,
+                                          self.points, edges)
+            renderer.render(
+                render_lines=False, line_colour='r', line_width=4,
```
Comment (patricksnape) [inline menpo/shape/pointcloud.py]: Are the line width and color supposed to be hardcoded here? Surely they aren't even used? Shouldn't we just ignore these arguments?

File: menpo/shape/mesh/textured.py [hunk 3]
```diff
@@ -38,6 +38,15 @@ def __init__(self, points, tcoords, texture, trilist=None, copy=True):
         else:
             self.texture = texture.copy()
 
+    @property
+    def n_channels(self):
```
Comment (patricksnape) [inline menpo/shape/mesh/textured.py]: `n_texture_channels`?

File: menpo/shape/mesh/coloured.py [hunk 4]
```diff
@@ -162,38 +171,169 @@ def from_mask(self, mask):
...
+            A copy of this mesh with colours linearly rescaled to fit in the
+            range provided.
+        """
+        instance = self.copy()
+        colours = instance.colours
+        if per_channel:
+            min_, max_ = colours.min(axis=0), colours.max(axis=0)
+        else:
+            min_, max_ = colours.min(), colours.max()
+        sf = ((maximum - minimum) * 1.0) / (max_ - min_)
```
Comment (patricksnape) [inline menpo/shape/mesh/coloured.py]: Why `* 1.0`?

File: menpo/shape/mesh/coloured.py [hunk 5]
```diff
@@ -127,6 +127,15 @@ def init_from_depth_image(cls, depth_image, colours=None):
                    trilist=new_tmesh.trilist,
                    copy=False)
 
+    @property
+    def n_channels(self):
```
Comment (patricksnape) [inline menpo/shape/mesh/coloured.py]: `n_colour_channels`?

File: menpo/shape/mesh/textured.py (hunk 3 above)
Comment (nontas) [inline menpo/shape/mesh/textured.py]: I don't see the point for this. The term `channels` is clear...

File: menpo/shape/mesh/textured.py (hunk 1 above)
Comment (jabooth) [inline menpo/shape/mesh/textured.py]: See https://github.com/menpo/menpo/pull/761#issuecomment-263526822

File: menpo/shape/pointcloud.py (hunk 2 above)
Comment (nontas) [inline menpo/shape/pointcloud.py]: True! I removed them

File: menpo/shape/mesh/coloured.py (hunk 5 above)
Comment (nontas) [inline menpo/shape/mesh/coloured.py]: Similar to `n_texture_channels`. I don't think it's necesary

File: menpo/shape/mesh/coloured.py (hunk 4 above)
Comment (nontas) [inline menpo/shape/mesh/coloured.py]: Yeah that's redundant! 

Btw should we rename this method to `rescale_texture()` in order to be in correspondence with `rescale_pixels()` of `Image`?

File: menpo/shape/mesh/textured.py (hunk 3 above)
Comment (patricksnape) [inline menpo/shape/mesh/textured.py]: Hmm, ok - it's not immediately clear to me that `n_channels` on a trimesh means the texture channels but I see the point about coloured meshes.

File: menpo/shape/mesh/coloured.py [hunk 6]
```diff
@@ -162,38 +171,170 @@ def from_mask(self, mask):
...
+        Returns
+        -------
+        self : :map:`ColouredTriMesh`
+            A copy of self with its texture clipped.
+        """
+        instance = self.copy()
+        instance.colours = np.clip(self.colours, *range)
+        return instance
+
+    def with_rescaled_texture(self, minimum, maximum, per_channel=True):
```
Comment (patricksnape) [inline menpo/shape/mesh/coloured.py]: Similarly here, we have `rescale_pixels` - so just `rescale_texture_pixels`?

File: menpo/shape/mesh/textured.py (hunk 1 above)
Comment (nontas) [inline menpo/shape/mesh/textured.py]: done

**Pull request merged**